pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java

```diff
@@ -152,39 +202,82 @@ private TransferableBlock buildJoinedDataBlock(TransferableBlock leftBlock)
     if (leftBlock.isErrorBlock()) {
       _upstreamErrorBlock = leftBlock;
       return _upstreamErrorBlock;
-    } else if (TransferableBlockUtils.isNoOpBlock(leftBlock) || TransferableBlockUtils.isEndOfStream(leftBlock)) {
+    }
+    if (leftBlock.isNoOpBlock() || (leftBlock.isSuccessfulEndOfStreamBlock() && !needUnmatchedRows())) {
       return leftBlock;
     }
+    if (leftBlock.isSuccessfulEndOfStreamBlock()) {
+      // Return remaining non-matched rows for non-inner join.
+      List<Object[]> returnRows = new ArrayList<>();
+      if (needUnmatchedRightRows()) {
+        for (Map.Entry<Key, List<Object[]>> entry : _broadcastRightTable.entrySet()) {
+          Set<Integer> matchedIdx = _matchedRightRows.getOrDefault(entry.getKey(), new HashSet<>());
+          int rowSize = entry.getValue().size();
+          Set<Integer> allIdx = IntStream.range(0, rowSize).boxed().collect(Collectors.toSet());
+          Set<Integer> unmatchedIdx = Sets.difference(allIdx, matchedIdx);
+          for (int unmatchedRow : unmatchedIdx) {
+            Object[] resultRow = joinRow(null, entry.getValue().get(unmatchedRow));
+            returnRows.add(resultRow);
+          }
```

Comment:
can we avoid using stream? this is on the critical code path (for every hashkey) so it is better to do a single loop

```suggestion
          Set<Integer> matchedIdx = _matchedRightRows.getOrDefault(entry.getKey(), new HashSet<>());
          List<Object[]> rightRows = entry.getValues();
          for (int i = 0; i <  rightRows.size(); i++) {
            if (!matchedIdx.contains(i)) {
              Object[] resultRow = joinRow(null, rightRows.get(i));
              returnRows.add(resultRow);
            }
          }
```

Reply:
Done